src/compiler/parser.ts

```diff
@@ -1922,20 +2000,25 @@ namespace ts {
         // TYPES
 
         function parseTypeReferenceOrTypePredicate(): TypeReferenceNode | TypePredicateNode {
-            let typeName = parseEntityName(/*allowReservedWords*/ false, Diagnostics.Type_expected);
+            contextualError.callback = invalidTypeReferenceOrTypePredicate;
+
+            let typeName = parseEntityName(/*allowReservedWords*/ false);
             if (typeName.kind === SyntaxKind.Identifier && token === SyntaxKind.IsKeyword && !scanner.hasPrecedingLineBreak()) {
                 nextToken();
-                let node = <TypePredicateNode>createNode(SyntaxKind.TypePredicate, typeName.pos);
-                node.parameterName = <Identifier>typeName;
-                node.type = parseType();
-                return finishNode(node);
+                let predicateNode = <TypePredicateNode>createNode(SyntaxKind.TypePredicate, typeName.pos);
+                predicateNode.parameterName = <Identifier>typeName;
+                predicateNode.type = parseType();
+                return finishNodeContextualError(predicateNode);
             }
+
             let node = <TypeReferenceNode>createNode(SyntaxKind.TypeReference, typeName.pos);
             node.typeName = typeName;
-            if (!scanner.hasPrecedingLineBreak() && token === SyntaxKind.LessThanToken) {
-                node.typeArguments = parseBracketedList(ParsingContext.TypeArguments, parseType, SyntaxKind.LessThanToken, SyntaxKind.GreaterThanToken);
+            if ((typeName.flags & NodeFlags.Missing) === 0) {

```

Comment:
As @CyrusNajmabadi mentioned, we already had a way of checking if something was missing, so we shouldn't introduce another.